jstz_api/src/encoding/text_decoder.rs

```diff
@@ -0,0 +1,381 @@
+        //  1. Let encoding be the result of getting an encoding from label.
+        //  2. If encoding is failure or replacement, then throw a RangeError.
+        let encoding =
+            encoding_rs::Encoding::for_label_no_replacement(label.as_bytes()).ok_or(())?;
+        Ok(TextDecoder {
+            //  3. Set this's encoding to encoding.
+            encoding: encoding,
+            //  4. If options["fatal"] is true, then set this's error mode to "fatal".
+            error_mode: if options.fatal {
+                "fatal".to_string()
+            } else {
+                "replacement".to_string()
+            },
+            decoder: if options.ignore_bom {
+                encoding.new_decoder_without_bom_handling()
+            } else {
+                encoding.new_decoder()
+            },
+            io_queue: Vec::new(),
+            //  5. Set this's ignore BOM to options["ignoreBOM"].
+            ignore_bom: options.ignore_bom,
+            //  A TextDecoder object has an associated do not flush, which is a boolean, initially false.
+            do_not_flush: false,
+        })
+    }
+    fn encoding(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextDecoder,
+            "encoding",
+            get:((x, context) => Ok(x.encoding.name().to_string().to_lowercase().into_js(context)))
+        )
+    }
+    fn fatal(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextDecoder,
+            "fatal",
+            get:((x, _context) => Ok((x.error_mode == "fatal").into()))
+        )
+    }
+    fn ignore_bom(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextDecoder,
+            "ignoreBOM",
+            get:((x, _context) => Ok(x.ignore_bom.into()))
+        )
+    }
+
+    //  https://encoding.spec.whatwg.org/#concept-td-serialize
+    fn serialize(&mut self, read: usize) -> () {
+        //  1. Let output be the empty string.
+        //  2. While true:
+        //    1. Let item be the result of reading from ioQueue.
+        //    2. If item is end-of-queue, then return output.
+        //    3. If decoder’s encoding is UTF-8 or UTF-16BE/LE, and decoder’s ignore BOM and BOM seen are false, then:
+        //      1. Set decoder’s BOM seen to true.
+        //      2. If item is U+FEFF, then continue.
+        //  4. Append item to output.
+
+        //  NB: we call this function in case of a chunked input with `read`:
+        //      https://docs.rs/encoding_rs/latest/encoding_rs/struct.Decoder.html
+        //  NB: BOM handling is done at the level of the decoding function from `::encoding_rs` crate
+        self.io_queue.drain(0..read);
+    }
+
+    //  https://encoding.spec.whatwg.org/#dom-textdecoder-decode
+    fn decode(
+        // input: Option<allowsharedbufersource>,
+        &mut self,
+        input: Option<Vec<u8>>,
```

Comment:
I'm surprised you'd have ownership of the `Vec<u8>`? I would've expected maybe `&[u8]` (default is empty) or `Option<&[u8]>`

Reply:
I tried to fight the borrow checker but TBH I couldn't achieve that.. sorry